### PR TITLE
Refactor of tokenization and detokenization within the NMT model

### PR DESCRIPTION
**Pull request by MaximumEntropy**
This PR refactors some of the fetching of tokenizers, normalizers, and detokenizers within the model and creates Chinese/Japanese tokenizer and detokenizer classes to match Moses semantics.

### CODE REVIEW

**Comment by mchrzanowski** [comment]
thank you for doing this. lgtm.

File: nemo/collections/common/tokenizers/japanese_tokenizers.py [hunk 1]
```diff
@@ -45,14 +45,15 @@ def __init__(self, sp_tokenizer_model):
     def sp_tokenize(self, text: str) -> str:
         return ' '.join(self.sp_tokenizer.text_to_tokens(text))
     
-    def tokenize(self, text):
+    def tokenize(self, text, escape=False, return_str=False):
         """
         Detokenizes a list of sentencepiece tokens in Japanese
```
Comment (RPrenger) [inline nemo/collections/common/tokenizers/japanese_tokenizers.py:53]: Nit:  The comment should be "Tokenizes text through moses then sentencepiece" or possibly removed entirely.

Comment (MaximumEntropy) [reply nemo/collections/common/tokenizers/japanese_tokenizers.py]: Fixed. Thanks :)

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py [hunk 2]
```diff
@@ -381,38 +407,16 @@ def translate(self, text: List[str], source_lang: str = None, target_lang: str =
...
 
         try:
             self.eval()
             res = []
             for txt in text:
                 if source_lang != "None":
-                    if source_lang == "ja":
-                        txt = normalizer.normalize(txt)
-                        txt = tokenizer1.tokenize(txt, escape=False, return_str=True)
-                        txt = ' '.join(self.sentencepiece_aux_tokenizer.text_to_tokens(txt))
```
Comment (okuchaiev) [inline nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:409]: this route for Japanese got removed?

**Review by okuchaiev — CHANGES_REQUESTED** [review]

File: nemo/collections/common/tokenizers/japanese_tokenizers.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class JapaneseDetokenizer:
+    def __init__(self):
+        self.moses_detokenizer = MosesDetokenizer(lang='ja')
+        self.sp_detokenizer = SentencePieceDetokenizer()
+
+    def detokenize(self, tokens: List[str]) -> str:
+        """
+        Detokenizes a list of sentencepiece tokens in Japanese
```
Comment (mchrzanowski) [inline nemo/collections/common/tokenizers/japanese_tokenizers.py:30]:
```suggestion
        Detokenizes a list of Japanese tokens
```

File: nemo/collections/common/tokenizers/japanese_tokenizers.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            detokenized Japanese string
+        """
+        text = self.sp_detokenizer.detokenize(tokens)
+        return self.moses_detokenizer.detokenize(text)
+
+
+class JapaneseTokenizer:
+    def __init__(self, sp_tokenizer_model):
+        self.moses_tokenizer = MosesTokenizer(lang='ja')
+        self.sp_tokenizer = SentencePieceTokenizer(model_path=sp_tokenizer_model)
```
Comment (mchrzanowski) [inline nemo/collections/common/tokenizers/japanese_tokenizers.py:43]: can you make sure to say that `sp_tokenizer_model` has to be a string to a sentencepiece model, not an actual sentencepiece object?

**Review by mchrzanowski — CHANGES_REQUESTED** [review]

File: nemo/collections/common/tokenizers/japanese_tokenizers.py (hunk 4 above)
Comment (MaximumEntropy) [reply nemo/collections/common/tokenizers/japanese_tokenizers.py]: Sure, I've added a docstring for the class and changed the variable name to `sp_tokenizer_model_path`. Thanks!

**Comment by mchrzanowski** [comment]
Lgtm. Thanks again!

**Review by mchrzanowski — APPROVED** [review]

**Review by mchrzanowski — APPROVED** [review]

**Review by okuchaiev — APPROVED** [review]

**Pull request merged**